File: tests/typesafety/typescript/logged_steps.ts
// expect: passes

type Logged<T> = { readonly value: T; readonly log: readonly string[] };

function andThen<T, U>(logged: Logged<T>, step: (value: T) => Logged<U>): Logged<U> {
  const result = step(logged.value);
  return { value: result.value, log: [...logged.log, ...result.log] };
}

function double(n: number): Logged<number> {
  return { value: n * 2, log: [`doubled ${n}`] };
}

function addTax(n: number): Logged<number> {
  return { value: n + Math.floor(n / 10), log: [`taxed ${n}`] };
}

export function total(n: number): Logged<number> {
  return andThen(andThen({ value: n, log: [] }, double), addTax);
}

File: tests/typesafety/python/logged_steps.py
# expect: passes
# run: yes
# title: Logged pairs each result with the log that produced it
# improves: printed_steps
"""`Logged` pairs a value with the log that produced it. `and_then` runs the
next step and concatenates the trails, so the log arrives with the answer,
as data. This shape is usually called the Writer monad."""

from collections.abc import Callable
from dataclasses import dataclass


@dataclass(frozen=True)
class Logged[T]:
    value: T
    log: tuple[str, ...]


def and_then[T, U](logged: Logged[T], step: Callable[[T], Logged[U]]) -> Logged[U]:
    result = step(logged.value)
    return Logged(result.value, logged.log + result.log)


def double(n: int) -> Logged[int]:
    return Logged(n * 2, (f"doubled {n}",))


def add_tax(n: int) -> Logged[int]:
    return Logged(n + n // 10, (f"taxed {n}",))


def total(n: int) -> Logged[int]:
    return and_then(and_then(Logged(n, ()), double), add_tax)


assert total(100).value == 220
assert total(100).log == ("doubled 100", "taxed 200")
